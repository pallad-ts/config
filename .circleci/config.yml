version: 2.1

orbs:
  node: circleci/node@4.7

jobs:
  main_job:
    executor:
      name: node/default
      tag: "14.18"
    steps:
      - checkout
      - run: "npm ci"
      - run:
          command: ./node_modules/.bin/lerna run compile
          name: compilation
      - run:
          command: ./node_modules/.bin/lerna run lint
          name: linter
      - run:
          command: ./node_modules/.bin/lerna run test
          name: test
      - store_test_results:
          path: ./test_results

workflows:
  main:
    jobs:
      - main_job
