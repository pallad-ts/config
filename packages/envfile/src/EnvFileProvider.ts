import { left, right } from "@sweet-monads/either";
import { fromNullable, none, just } from "@sweet-monads/maybe";

import { Provider } from "@pallad/config";
import { ValueNotAvailable } from "@pallad/config";

export class EnvFileProvider extends Provider<string> {
    constructor(
        private key: string,
        private envFileConfigLoader: () => Record<string, string>
    ) {
        super();
    }

    getValue() {
        const config = this.envFileConfigLoader();
        return fromNullable(config[this.key])
            .chain(value => (value === "" ? none() : just(value)))
            .fold(
                () => {
                    return left(new ValueNotAvailable(`"${this.key}" from ENV file(s)`));
                },
                result => {
                    return right(result);
                }
            );
    }
}
