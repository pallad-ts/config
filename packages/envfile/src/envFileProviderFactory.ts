import { parse } from "dotenv";
import { readFileSync, existsSync } from "fs";
import { resolve } from "path";

import { runOnlyOnceWrapper } from "@pallad/config-common";

import { EnvFileProvider } from "./EnvFileProvider";
import { ERRORS } from "./errors";

export function envFileProviderFactory(options: envFileProviderFactory.Options) {
    function loadEnvFiles() {
        const envs: Record<string, string> = {};

        const cwd = options?.cwd ?? process.cwd();

        const paths = Array.isArray(options.paths) ? options.paths : [options.paths];

        for (const path of paths) {
            const resolvedPathData: envFileProviderFactory.PathEntryObject =
                typeof path === "string" ? { path: path, required: true } : path;

            const filePath = resolve(cwd, resolvedPathData.path);

            if (!existsSync(filePath)) {
                if (resolvedPathData.required) {
                    throw ERRORS.ENV_FILE_DOES_NOT_EXIST.create(resolvedPathData.path);
                }
                continue;
            }

            const result = parse(readFileSync(filePath, "utf-8"));
            Object.assign(envs, result);
        }

        if (options?.populateToEnv) {
            if (typeof options.populateToEnv === "function") {
                for (const [key, value] of Object.entries(envs)) {
                    if (options.populateToEnv(key, value)) {
                        process.env[key] = value;
                    }
                }
            } else if (options.populateToEnv) {
                Object.assign(process.env, envs);
            }
        }

        return envs;
    }

    return (key: string) => new EnvFileProvider(key, runOnlyOnceWrapper(loadEnvFiles));
}

export namespace envFileProviderFactory {
    export interface Options {
        paths: PathEntry | PathEntry[];
        /**
         * Working directory. By default `process.cwd()`
         */
        cwd?: string;
        /**
         * Whether to populate loaded values to process.env. Default: false
         */
        populateToEnv?: boolean | PopulateToEnvPredicate;
    }

    export type PathEntry = string | PathEntryObject;

    export interface PathEntryObject {
        path: string;
        required: boolean;
    }

    export interface PopulateToEnvPredicate {
        (key: string, value: string): boolean;
    }
}
