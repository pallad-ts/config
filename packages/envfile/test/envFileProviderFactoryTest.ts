import { EnvFileProvider } from "@src/EnvFileProvider";
import { envFileProviderFactory } from "@src/envFileProviderFactory";
import { left, right } from "@sweet-monads/either";
import { assert, IsExact } from "conditional-type-checks";
import * as sinon from "sinon";

import { DefaultValueProvider, ValueNotAvailable } from "@pallad/config";

describe("envFileProviderFactory", () => {
    describe("types", () => {
        const factory = envFileProviderFactory({
            paths: "./example/base.env",
            cwd: __dirname,
        });

        it("no default and transformer", () => {
            const provider = factory("FOO");

            type Input = typeof provider;
            type Expected = EnvFileProvider;
            assert<IsExact<Input, Expected>>(true);
        });
    });

    it("simple case", () => {
        const factory = envFileProviderFactory({
            paths: "./example/base.env",
            cwd: __dirname,
        });

        expect(factory("FOO").getValue()).toStrictEqual(right("bar"));

        expect(factory("HELLO").getValue()).toStrictEqual(right("world"));
    });

    it("overriding env from previous files", () => {
        const factory = envFileProviderFactory({
            paths: ["./example/base.env", "./example/override.env"],
            cwd: __dirname,
        });

        expect(factory("FOO").getValue()).toStrictEqual(right("world"));
    });

    it("fails if env file does not exist and ignoring is disabled", () => {
        expect(() => {
            envFileProviderFactory({
                paths: "./example/test.env",
                cwd: __dirname,
            }).load();
        }).toThrowErrorMatchingSnapshot();
    });

    it("silently ignores file if it is not required", () => {
        const factory = envFileProviderFactory({
            paths: { path: "./example/test.env", required: false },
            cwd: __dirname,
        });

        expect(factory("FOO").getValue()).toStrictEqual(left(new ValueNotAvailable('"FOO" from ENV file(s)')));
    });

    describe("populating to env", () => {
        let originalEnv = process.env;
        beforeEach(() => {
            originalEnv = { ...process.env };
        });

        afterEach(() => {
            process.env = { ...originalEnv };
        });

        it("all env", () => {
            const factory = envFileProviderFactory({
                paths: "./example/populate.env",
                cwd: __dirname,
            });

            factory.populateToEnv();
            expect(process.env.PALLAD_CONFIG_TEST_VALUE).toStrictEqual("test");
            expect(process.env.PALLAD_CONFIG_TEST_VALUE2).toStrictEqual("test2");
        });

        it("only the ones accepted by predicate", () => {
            const factory = envFileProviderFactory({
                paths: "./example/populate.env",
                cwd: __dirname,
            });

            const stub = sinon.stub().callsFake((key: string) => key === "PALLAD_CONFIG_TEST_VALUE2");
            factory.populateToEnv(stub);

            expect(process.env.PALLAD_CONFIG_TEST_VALUE).toBeUndefined();
            expect(process.env.PALLAD_CONFIG_TEST_VALUE2).toStrictEqual("test2");

            sinon.assert.calledWith(stub, "PALLAD_CONFIG_TEST_VALUE", "test");
            sinon.assert.calledWith(stub, "PALLAD_CONFIG_TEST_VALUE2", "test2");
        });
    });
});
