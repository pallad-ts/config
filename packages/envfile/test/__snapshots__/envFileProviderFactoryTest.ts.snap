// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`envFileProviderFactory fails if env file does not exist and ignoring is disabled 1`] = `"Env file "./example/test.env" does not exist but it is required"`;
